core/src/main/scala/akka/kafka/internal/TransactionalSource.scala

```diff
@@ -36,9 +81,73 @@ private[kafka] final class TransactionalSource[K, V](consumerSettings: ConsumerS
     IsolationLevel.READ_COMMITTED.toString.toLowerCase(Locale.ENGLISH)
   )
 
+  case object Drained
+  case class Drain[T](ack: ActorRef, msg: T)
+  case class Committed(offsets: Map[TopicPartition, OffsetAndMetadata])
+
   override protected def logic(shape: SourceShape[TransactionalMessage[K, V]]): GraphStageLogic with Control =
     new SingleSourceLogic[K, V, TransactionalMessage[K, V]](shape, txConsumerSettings, subscription)
     with TransactionalMessageBuilder[K, V] {
+      var inFlightRecords = InFlightRecords.empty
+
+      override def messageHandling = super.messageHandling.orElse(drainHandling).orElse {
+        case (_, Revoked(tps)) =>
+          inFlightRecords.revoke(tps.toSet)
+      }
+
+      override def shuttingDownReceive = super.shuttingDownReceive.orElse(drainHandling)
+
+      def drainHandling: PartialFunction[(ActorRef, Any), Unit] = {
+        case (sender, Committed(offsets)) =>
+          inFlightRecords.committed(offsets.mapValues(_.offset() - 1))
+          sender ! Done
+        case (_, Drain(ack, msg)) =>
+          if (inFlightRecords.empty()) {
+            log.debug("Source drained")
+            ack ! msg
+          } else {
+            log.debug(s"Draining partitions {}", inFlightRecords)
+            materializer.scheduleOnce(10.millis, new Runnable {
+              override def run(): Unit =
+                sourceActor.ref ! Drain(ack, msg)
+            })
+          }
+      }
+
       override def groupId: String = txConsumerSettings.properties(ConsumerConfig.GROUP_ID_CONFIG)
+      lazy val committedMarker: CommittedMarker = {
+        val ec = materializer.executionContext
+        CommittedMarkerRef(sourceActor.ref, commitTimeout)(ec)
+      }
+
+      override def onMessage(rec: ConsumerRecord[K, V]): Unit =
+        inFlightRecords.add(Map(new TopicPartition(rec.topic(), rec.partition()) -> rec.offset()))
+
+      override protected def stopConsumerActor(): Unit =
+        sourceActor.ref.tell(Drain(consumerActor, KafkaConsumerActor.Internal.Stop), sourceActor.ref)
+
+      // This is invoked in the KafkaConsumerActor thread when doing poll.
+      override def partitionRevokedCB(revokedTps: Set[TopicPartition]): Unit = {
+        waitForDraining()
+        sourceActor.ref ! Revoked(revokedTps.toList)
+        super.partitionRevokedCB(revokedTps)
+      }
+
+      def waitForDraining(): Unit = {
+        import akka.pattern.ask
+        implicit val timeout = Timeout(5.seconds)
+        Await.result(ask(stageActor.ref, Drain(stageActor.ref, Drained)), 10.seconds)
```

Comment:
I'm afraid if the behaviour in case of a timeout is correct.
IIUC TimeoutException is thrown and it's caught in kafka code by ConsumerCoordinator,
here:

```
try {
            listener.onPartitionsRevoked(revoked);
        } catch (WakeupException | InterruptException e) {
            throw e;
        } catch (Exception e) {
            log.error("User provided listener {} failed on partition revocation", listener.getClass().getName(), e);
        }
```

So due to `catch (Exception e)` we won't fail the stream and we'll just continue like nothing happened and the stream was drained.

Therefore if it takes too long to drain then we still allow messages to remain in the stream. In the meantime some other consumer might start processing data from the same partition because we just released it.

@szymonm @2m Could you verify if my reasoning is correct?

Reply:
You're right @jotek7. In this case we should fail the stream starting from the producer. Which is similar to what we should do generally in the failure case (see my comments)